blog/2024-11-19-microsoft-azure-authentication.md

```diff
@@ -0,0 +1,348 @@
+---
+author: Tom Gotsman
+date: 2024-11-19
+title: Reflex Authentication with Microsoft Azure
+description: A practical guide to implementing Microsoft Azure Single Sign-On (SSO) authentication in Reflex applications.
+image: /blog/self-hosting-with-docker.webp
+meta: [
+    {
+      "name": "keywords",
+      "content": "
+        data-driven web apps,
+        modern web application development,
+        python data visualization,
+        python fintech applications,
+        python web development,
+      "
+    },
+]
+---
+
+
+```python exec
+from pcweb.pages import docs
+```
+
+Authentication is a critical requirement for most internal company web applications. While building features and functionality is exciting, ensuring secure access control often becomes a necessary next step. For organizations already using Microsoft's ecosystem, this blog post will guide you through setting up Reflex authentication (SSO) with Microsoft Azure.
+
+
+## Variables to Set
+
+Before we begin, you will need the following imports and to set the following variables:
+
+```python
+import msal
+import reflex as rx
+from typing import Dict, List, Optional
+
+
+client_id: str = "0df2a88e-####-####-####-#########373"
+client_secret: str = "TPO8Q~dlz###########################"
+tenant_id: str = "f2c9cbbe-####-####-####-#########d6d"
+
+# This tells which company is issuing your credentials
+authority = f"https://login.microsoftonline.com/\{tenant_id}"
+# Customizable to where the user after login will get re-directed
+login_redirect = "/"
+# Creates an in-memory cache for storing tokens
+cache = msal.TokenCache()
+```
+
+The Microsoft Authentication Library (MSAL) for Python library enables you to sign in users or apps with Microsoft identities. 
+
+The values you should get from your Azure portal / SSO team at your company are `client_id`, `client_secret`, and `tenant_id`. These values are unique to your application and company.
+
+It is recommended to retrive these values from environment variables or from a configuration file (they are just hardcoded for the example for simplicity).
+
+Next we have to set the `sso_app` variable, which is the client application that will be used to authenticate users. It is strongly recommended to set a `client_secret` for production applications.
+
+```python
+sso_app: msal.ClientApplication
+if client_secret:
+    sso_app = msal.ConfidentialClientApplication(
+        client_id=client_id,
+        client_credential=client_secret,
+        authority=authority,
+        token_cache=cache,
+    )
+else:
+    sso_app = msal.PublicClientApplication(
+        client_id=client_id,
+        authority=authority,
+        token_cache=cache,
+    )
+```
```

Comment:
Maybe we should only show them the `ConfidentialClientApplication` since in the variables above we set the `client_secret`

and mention that there is a `PublicClientApplication` that does not require `client_secret` but is not recommended until you know what you are doing ?

Reply:
will remove in the blog